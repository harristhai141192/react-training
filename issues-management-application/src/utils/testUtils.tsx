// Libraries
import React from 'react';
import { ThemeProvider } from '@chakra-ui/react';
import { render } from '@testing-library/react';
import { BrowserRouter as Router } from 'react-router-dom';

// Themes
import themes from '@themes/index';

const customRender = (component: React.ReactNode) => {
  return render(
    <ThemeProvider theme={themes}>
      <Router>{component}</Router>
    </ThemeProvider>,
  );
};

export * from '@testing-library/react';
export { customRender as render };
