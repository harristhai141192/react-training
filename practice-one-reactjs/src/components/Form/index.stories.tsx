// Libraries
import { ComponentMeta, ComponentStory } from '@storybook/react';

// Components
import Form from '.';

// CSS
import '../../index.css';

export default {
  title: 'Practice-One/Form',
  component: Form,
} as ComponentMeta<typeof Form>;

const Template: ComponentStory<typeof Form> = (args) => <Form {...args} />;

export const Primary = Template.bind({});

Primary.args = {
  formTitle: 'Add New Pokemon',
  pokemonData: {},
};
