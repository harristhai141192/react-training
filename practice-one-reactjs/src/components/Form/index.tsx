// Libraries
import React, { ChangeEvent, FormEvent, useState } from 'react';

// Components
import Button from '@components/Button';
import Input from '@components/Input';
import Select from '@components/SelectInput';
import TextField from '@components/TextField';
import { IPokemonProps } from '@models/pokemon';
import { POKEMON_ELEMENTS, POKEMON_TYPE } from '@constants/variables';

// CSS
import './index.styles.css';
import Link from '@components/Link';

interface IProps {
  pokemonData: IPokemonProps;
  formTitle?: string;
  isEdit?: boolean;
  onFormSubmit: (e: FormEvent<HTMLFormElement>) => void;
  handleOnChange: (e: ChangeEvent<HTMLInputElement>) => void;
  handleOnChangeSelect: (e: ChangeEvent<HTMLSelectElement>) => void;
  handleOnChangeTextArea: (e: ChangeEvent<HTMLTextAreaElement>) => void;
  onFormCancel?: () => void;
}

const Form: React.FC<IProps> = ({
  formTitle,
  pokemonData,
  isEdit,
  onFormSubmit,
  handleOnChange,
  handleOnChangeSelect,
  handleOnChangeTextArea,
}) => {
  return (
    <>
      <h3 className='formTitle'>{formTitle}</h3>
      <form onSubmit={onFormSubmit}>
        <Input
          name='name'
          value={pokemonData.name || ''}
          labelName='Pokemon Name'
          error={false}
          required={true}
          onChange={handleOnChange}
        />
        <Input
          value={pokemonData.code || ''}
          name='code'
          labelName='Pokemon Number'
          type='number'
          error={false}
          required={true}
          onChange={handleOnChange}
        />
        <Input
          value={pokemonData.photo || ''}
          name='photo'
          labelName='Picture'
          type='text'
          error={false}
          onChange={handleOnChange}
        />
        <Select
          listOption={POKEMON_ELEMENTS}
          name='element'
          labelName='Type 1'
          required={true}
          onChange={handleOnChangeSelect}
        />
        <Select
          listOption={POKEMON_TYPE}
          name='type2'
          labelName='Type 2'
          required={true}
          onChange={handleOnChangeSelect}
        />
        <TextField
          value={pokemonData.description}
          labelName='pokemonDescription'
          name='description'
          onChange={handleOnChangeTextArea}
          rows={2}
        />
        <Button type='submit' style='primary' size='large' label='Save' />
        <Link href={isEdit ? '/edit' : '/'}>
          <Button style='outlined' size='large' label='Cancel' type='button' />
        </Link>
      </form>
    </>
  );
};

export default Form;
