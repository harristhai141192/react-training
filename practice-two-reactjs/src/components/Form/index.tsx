// Libraries
import { FormControl, FormLabel, Box, Text, Textarea, Select } from '@chakra-ui/react';
//  IMPORT REACT NEN DE TREN CUNG
import React, { FormEvent } from 'react';

// Components
import InputComponent from '@components/Input';
import Button from '@components/Button';
import { IMemberDetail } from '@models/index';

// ERROR MESSAGE
interface IProps {
  defaultMemberData?: IMemberDetail;
  formName?: string;
  errorName?: boolean;
  errorDOB?: boolean;
  errorPhone?: boolean;
  errorMember?: boolean;
  errorEmail?: boolean;
  isLoading?: boolean;
  onSubmit?: (e: FormEvent<HTMLFormElement>) => void;
  onCancel?: React.MouseEventHandler<HTMLButtonElement>;
}

const FormComponent: React.FC<IProps> = ({
  defaultMemberData,
  formName,
  errorName,
  errorDOB,
  errorPhone,
  errorMember,
  errorEmail,
  isLoading,
  onSubmit = () => {},
  onCancel = () => {},
}) => {
  return (
    <form onSubmit={onSubmit}>
      <Box display='flex' flexDirection='column'>
        <Text fontSize='2xl'>{formName}</Text>
        <FormControl>
          <FormLabel marginTop={3} htmlFor='name' fontSize='sm'>
            Your name:
          </FormLabel>
          <InputComponent
            name='name'
            error={errorName}
            defaultValue={defaultMemberData?.memberName}
          />
          <FormLabel marginTop={3} htmlFor='dateOfBirth' fontSize='sm'>
            Date of birth:
          </FormLabel>
          <InputComponent
            name='dateOfBirth'
            type='date'
            error={errorDOB}
            defaultValue={defaultMemberData?.dateOfBirth}
          />
          <FormLabel marginTop={3} htmlFor='memberImage' fontSize='sm'>
            Member Image:
          </FormLabel>
          <InputComponent name='memberImage' defaultValue={defaultMemberData?.memberImg} />
          <FormLabel marginTop={3} htmlFor='phone' fontSize='sm'>
            Phone:
          </FormLabel>
          <InputComponent name='phone' error={errorPhone} defaultValue={defaultMemberData?.phone} />
          <FormLabel marginTop={3} htmlFor='memberSince' fontSize='sm'>
            Member since:
          </FormLabel>
          <InputComponent
            name='memberSince'
            type='date'
            error={errorMember}
            defaultValue={defaultMemberData?.memberSince}
          />
          <FormLabel marginTop={3} htmlFor='email' fontSize='sm'>
            Email:
          </FormLabel>
          <InputComponent
            name='gender'
            type='text'
            error={errorEmail}
            defaultValue={defaultMemberData?.gender}
          />
          <FormLabel marginTop={3} htmlFor='gender' fontSize='sm'>
            Gender:
          </FormLabel>
          <Select name='gender' marginTop={3} size='sm' defaultValue={defaultMemberData?.gender}>
            <option value='Male'>Male</option>
            <option value='Female'>Female</option>
          </Select>
          <FormLabel marginTop={3} htmlFor='job' fontSize='sm'>
            Job:
          </FormLabel>
          <InputComponent name='job' defaultValue={defaultMemberData?.job} />
          <FormLabel marginTop={3} htmlFor='description' fontSize='sm'>
            Description:
          </FormLabel>
          <Textarea
            marginTop={3}
            name='description'
            defaultValue={defaultMemberData?.description}
          ></Textarea>
          <Box
            marginTop={7}
            marginBottom={5}
            display='flex'
            justifyContent='flex-start'
            alignItems='flex-end'
          >
            <Button isLoading={isLoading} label='Submit' type='submit' loadingText='Submitting' />
            <Button label='Cancel' onClick={onCancel} isLoading={isLoading} />
          </Box>
        </FormControl>
      </Box>
    </form>
  );
};

export default FormComponent;
