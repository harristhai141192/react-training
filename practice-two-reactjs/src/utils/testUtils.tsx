import { ThemeProvider } from '@chakra-ui/react';
import { render } from '@testing-library/react';
import React from 'react';
import themes from '@themes/index';
import { BrowserRouter as Router } from 'react-router-dom';

const customRender = (component: React.ReactNode) => {
  return render(
    <ThemeProvider theme={themes}>
      <Router>{component}</Router>
    </ThemeProvider>,
  );
};

export * from '@testing-library/react';
export { customRender as render };
