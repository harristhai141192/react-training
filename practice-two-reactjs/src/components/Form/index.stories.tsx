// Libs
import { ComponentStory, ComponentMeta } from '@storybook/react';

// Component
import Form from '.';

export default {
  title: 'Components/Form',
  component: Form,
  argTypes: {},
} as ComponentMeta<typeof Form>;

const Template: ComponentStory<typeof Form> = (args) => <Form {...args} />;

export const Default = Template.bind({});
Default.args = {
  errorName: false,
  errorDOB: false,
  errorPhone: false,
  errorMember: false,
  errorEmail: true,
  errorEmailText: 'Please input correct email',
};
