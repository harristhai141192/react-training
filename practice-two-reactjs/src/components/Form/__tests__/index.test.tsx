// libraries
import { render, getByText } from '@utils/testUtils';
// components
import Form from '../index';
import '@testing-library/jest-dom';

// theme

describe('Component [Input] should match snapshot', () => {
  test('It should match snapshot', () => {
    const { container } = render(
      <Form
        errorName={false}
        errorDOB={false}
        errorPhone={false}
        errorMember={false}
        errorEmail={false}
        formName={''}
      />,
    );

    expect(container).toMatchSnapshot();
  });

  test('Component [Input] should render correctly', () => {
    const { container } = render(
      <Form
        errorName={false}
        errorDOB={false}
        errorPhone={false}
        errorMember={false}
        errorEmail={false}
        formName={'Add Form'}
      />,
    );
    const getTitle = getByText(container, 'Add Form');
    expect(getTitle).toBeInTheDocument();
  });
});
