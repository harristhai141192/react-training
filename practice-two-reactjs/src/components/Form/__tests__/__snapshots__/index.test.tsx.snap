// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component [Input] should match snapshot It should match snapshot 1`] = `
<div>
  <form>
    <div
      class="css-j7qwjs"
    >
      <p
        class="chakra-text css-9hpz1u"
      />
      <div
        class="chakra-form-control css-1kxonj9"
        role="group"
      >
        <label
          class="chakra-form__label css-g6pte"
          for="name"
          id="field-:r0:-label"
        >
          Your name: 
        </label>
        <input
          class="chakra-input css-10eaxd9"
          id="field-:r0:"
          name="name"
          placeholder=""
          value=""
        />
        <label
          class="chakra-form__label css-g6pte"
          for="dateOfBirth"
          id="field-:r0:-label"
        >
          Date of birth:
        </label>
        <input
          class="chakra-input css-10eaxd9"
          id="field-:r0:"
          name="dateOfBirth"
          placeholder=""
          value=""
        />
        <label
          class="chakra-form__label css-g6pte"
          for="phone"
          id="field-:r0:-label"
        >
          Phone:
        </label>
        <input
          class="chakra-input css-10eaxd9"
          id="field-:r0:"
          name="phone"
          placeholder=""
          value=""
        />
        <label
          class="chakra-form__label css-g6pte"
          for="memberDate"
          id="field-:r0:-label"
        >
          Member since:
        </label>
        <input
          class="chakra-input css-10eaxd9"
          id="field-:r0:"
          name="memberDate"
          placeholder=""
          type="date"
          value=""
        />
        <label
          class="chakra-form__label css-g6pte"
          for="email"
          id="field-:r0:-label"
        >
          Email:
        </label>
        <input
          class="chakra-input css-10eaxd9"
          id="field-:r0:"
          name="email"
          placeholder=""
          type="text"
          value=""
        />
        <div
          class="css-h2h2zh"
        >
          <button
            class="chakra-button css-jcr1fc"
            type="button"
          >
            Submit
          </button>
          <button
            class="chakra-button css-jcr1fc"
            type="button"
          >
            Cancel
          </button>
        </div>
      </div>
    </div>
  </form>
</div>
`;
